src/js/protocols/webstm32.js

```diff
@@ -0,0 +1,941 @@
+            go: 0x21, // Jumps to user application code located in the internal Flash memory or in SRAM
+            write_memory: 0x31, // Writes up to 256 bytes to the RAM or Flash memory starting from an address specified by the application
+            erase: 0x43, // Erases from one to all the Flash memory pages
+            extended_erase: 0x44, // Erases from one to all the Flash memory pages using two byte addressing mode (v3.0+ usart).
+            write_protect: 0x63, // Enables the write protection for some sectors
+            write_unprotect: 0x73, // Disables the write protection for all Flash memory sectors
+            readout_protect: 0x82, // Enables the read protection
+            readout_unprotect: 0x92, // Disables the read protection
+        };
+
+        // Erase (x043) and Extended Erase (0x44) are exclusive. A device may support either the Erase command or the Extended Erase command but not both.
+        this.available_flash_size = 0;
+        this.page_size = 0;
+        this.useExtendedErase = false;
+
+    }
+    // no input parameters
+    connect(port, baud, hex, options, callback) {
+        this.hex = hex;
+        this.port = port;
+        this.baud = baud;
+        this.callback = callback;
+
+        // we will crunch the options here since doing it inside initialization routine would be too late
+        this.options = {
+            no_reboot: false,
+            reboot_baud: false,
+            erase_chip: false,
+        };
+
+        if (options.no_reboot) {
+            this.options.no_reboot = true;
+        } else {
+            this.options.reboot_baud = options.reboot_baud;
+        }
+
+        if (options.erase_chip) {
+            this.options.erase_chip = true;
+        }
+
+        if (this.options.no_reboot) {
+            // TODO: update to use web serial / USB API
+            serial.connect(port, { baudRate: this.baud, parityBit: 'even', stopBits: 'one' });
+        } else {
+
+            let rebootMode = 0; // FIRMWARE
+
+            const startFlashing = () => {
+                if (rebootMode === 0) {
+                    return;
+                }
+
+                // refresh device list
+                if (PortHandler.dfuAvailable) {
+                    DFU.connect(hex, options);
+                } else {
+                    // TODO: update to use web serial / USB API
+                    prepareSerialPort();
+                    console.log('Connecting to serial port', this.port, this.baud);
+                    serial.connect(this.port, { bitrate: this.baud, parityBit: 'even', stopBits: 'one' }, function (openInfo) {
+                        if (openInfo) {
+                            this.initialize();
+                        } else {
+                            GUI.connect_lock = false;
+                            gui_log(i18n.getMessage('serialPortOpenFail'));
+                        }
+                    });
+                }
+            };
```

Comment:
Stylistic change, but could we move this outside of the class Or create a method instead?

It just becomes very hard to follow when we define many nested functions and then use them in the same method.

Reply:
Created a method